go_modules/helpers/updatechecker/main.go

```diff
@@ -29,7 +29,8 @@ func GetVersions(args *Args) (interface{}, error) {
 
 	currentVersion := args.Dependency.Version
 
-	modload.LoadModFile(context.Background())
+	modload.DisallowWriteGoMod()
+	_ = modload.LoadModFile(context.Background())
```

Comment:
This adapts to the signature change mentioned in https://github.com/dependabot/gomodules-extracted/pull/12#discussion_r691813720

I added a call `modload.DisallowWriteGoMod()` to keep the go 1.16 behavior of not writing back to the `go.mod` file. This doesn't really matter since we are only looking for new versions of a module here and not performing any updates yet. This runs with it's own copy of `go.mod` in a temp directory so any writes would get discarded: https://github.com/dependabot/dependabot-core/blob/b5cfb4970368a3f9c46b573de257b58bd4350961/go_modules/lib/dependabot/go_modules/update_checker/latest_version_finder.rb#L74-L75

The new return value of `modload.LoadModFile(context.Background())` didn't seem particularly useful so I adjusted the call to show we don't care about it.

re: https://github.com/dependabot/gomodules-extracted/pull/12#discussion_r691813491
I haven't found a reason to switch to `modload.LoadModGraph` here yet. It seems like it would do more work which wouldn't be relevant to the goal of checking for versions of the requested dependency.

I would like to revisit https://github.com/dependabot/dependabot-core/pull/3630 now that go 1.17 is out to see if it resolved the performance concerns we saw before. Switching to `go list` would remove the need to keep poking around with these go mod internal libs.

Reply:
/cc @jeffwidman Would appreciate your thoughts here as well!

Reply:
> I added a call modload.DisallowWriteGoMod() to keep the go 1.16 behavior of not writing back to the go.mod file.

👍  makes sense to me.

> The new return value of `modload.LoadModFile(context.Background())` didn't seem particularly useful so I adjusted the call to show we don't care about it.

👍 

> I would like to revisit #3630 now that go 1.17 is out to see if it resolved the performance concerns we saw before. Switching to go list would remove the need to keep poking around with these go mod internal libs.

Completely agree. Everything I wrote in https://github.com/dependabot/dependabot-core/pull/3630#issuecomment-833779506 still stands. It's been a few months since I looked deeply into the guts of `go` `1.17`'s new design for handling transitive deps, so I don't recall whether "in theory" it will now work or not. But what matters is "in practice" and that should be easy to test. Maybe re-open that #3630 PR so that there's a floating todo to look at it more?

🤞 it works, as it would simplify life greatly.